File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
      rotate_array(nums,k);
  }
   public static void rotate_array(int[] arr,int k){
        int n=arr.length;
        k=k%n;
        reverse(arr,0,n-k-1);//first k element
        reverse(arr,n-k,n-1);//last k element;
        reverse(arr,0,n-1);//reverse entire element
    }
    public static int[] reverse(int[] arr,int start,int end){
        int i=start;
        int j=end;
        while(i<j){
            int temp=arr[i];
            arr[i]=arr[j];
            arr[j]=temp;
            i++;
            j--;
        }
        return arr;

        
    }
}

File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n=nums.size();
        k=k%n;

        while(k--!=0){
            int temp=nums[n-1];
            for(int i=n-1;i>0;i--){
                nums[i]=nums[i-1];
            }
            nums[0]=temp;
        }
    }
};